dBFT: get rid of snapshot dependency

AnnaShaleva opened the pull request:
Close #86. Close #51. Close #58.

- [x] Decide about Header's Coinbase meaning.
- [x] Implement Coinbase verification.
- [x] Remove the rest of snapshot mentions.
- [x] Ensure that code works when unprocessed blocks are in the chain.
- [x] Refactor dBFT API which is closely based on snapshot.
- [x] Format code and improve comments.

AnnaShaleva committed:
dbft: set Coinbase to primary address of the current dBFT round

A part of #86.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: allow to apply Eth-based block verification for unsigned block

To be able to properly verify unsigned block after PrepareRequest receival
in WithVerifyBlock dBFT callback we need to ensure that (*DBFT).Author can
handle such unsugned block. Starting from the previous commit, we define
block's Coinbase as the Primary address for the dBFT round when the block
was accepted. And thus, return proper block author for proper verification.

The problem itself is described in https://github.com/bane-labs/go-ethereum/pull/87#issuecomment-1888656405.
Also ref. #51.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: adjust header difficulty calculation algorithm

The following rule is the most close to header's Difficulty definition
from dBFT's POW:
* If block's Primary is the first speaker of the consensus process for the
  corresponding block height, then it's "in-turn" difficulty.
* If block's Primary is some other node (i.e. CV happened during consensus
  process), then it's "out-of-turn" difficulty.

A part of #51. A major part of #58.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: enable header's difficulty checks

Now we have proper relations between difficulty and dBFT rounds, and thus
enable difficulty checks during block verifications.

A part of #51. Closes #58.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: remove useless snapshot mentions

A part of #51. This commit tides dBFT off the snapshot. Starting from this
commit DBFT engine perform its operations without snapshot creation.

A part of #51.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: use header difficulty stub for proper sealing tasks routing

Beacon chain uses block's difficulty to distinguish PoA and PoW engines.
Empty difficulty leads to skipped sealing tasks.

A part of #51.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented on consensus/dbft/dbft.go:
```diff
@@ -411,6 +407,15 @@ func New(config *params.DBFTConfig, db ethdb.Database) (*DBFT, error) {
+			// custom address corresponding to every consensus node to be set as Coinbase
+			// or it's OK to have the consensus node address itself be set as Coinbase?
+			// Currently, the second option is implemented.
+			h.Coinbase = ctx.Validators[ctx.PrimaryIndex].(*PublicKey).Account
```
@roman-khimov, @chenquanyu, need your opinion on this topic. We need to decide about Primary and Coinbase relation.

chenquanyu replied:
> @roman-khimov, @chenquanyu, need your opinion on this topic. We need to decide about Primary and Coinbase relation.

I think they should be the same, they are the miner of block.

AnnaShaleva commented on consensus/dbft/dbft.go:
```diff
@@ -629,13 +634,29 @@ func (c *DBFT) postBlock(b *types.Block) {
+	//
+	// Currently option 1 is implemented, but without enforced miner.etherbase check. Please,
+	// write your thoughts on this.
+	return header.Coinbase, nil
```
And this one is also closely related to the same Coinbase-Primary problem.

AnnaShaleva committed:
dbft: make header's Coinbase field be configurable

Make header's Coinbase be configurable via miner.etherbase node setting.
This commit allows Coinbase be different from the validator's address to be
able to send reward to a side account, e.g. to some Governance contract.
Validator-related coinbase check is removed as obsolete. This change is
needed to support NeoX governance model.

It is guaranteed by miner's constructor for Coinbase not to be empty. No
other validity checks is performed.

A part of #86.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: implement (*DBFT).Primary method

(*DBFT).Primary allows to retrieve the primary node address for the
speciied block. In common case Primary may differ from Author.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: align dBFT RPC API wrt the DBFT engine implementation

Make the dBFT RPC API follow the DBFT consensus engine implementation.
No setters are exposed since dBFT algorithm doesn't assume that. Instead,
provide some useful getters.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
miner: skip dBFT-related tests

Will be fixed in #95, and for now let's have non-failing test set
for future Eth core development.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@roman-khimov, ready for review. Coinbase topic discussion is moved to Discord, the second option (customizable Coinbase via `miner.etherbase`) is implemented in this PR.

AnnaShaleva commented on consensus/dbft/dbft.go:
```diff
@@ -771,94 +739,12 @@ func (c *DBFT) verifyCascadingFields(chain consensus.ChainHeaderReader, header *
-	c.recents.Add(snap.Hash, snap)
+	// TODO: ensure Coinbase is valid. If Coinbase is a Governance (or some other) contract, then ensure that
+	// Coinbase matches the contract address. If it's a validator address, then ensure Coinbase matches the
+	// Speaker address for the corresponding PrimaryIndex.
```
Once we decide on Coinbase meaning, I'll add Coinbase verification commit to this PR.

roman-khimov replied:
Looks like we have an agreement on this.

AnnaShaleva replied:
Fixed.

nicolegys commented:
Better not return error when getSigners of genesis block
![1705380185607](https://github.com/bane-labs/go-ethereum/assets/43694095/01841240-b87d-49f1-8f62-0d811adc28ad)

AnnaShaleva commented:
> Better not return error when getSigners of genesis block

@nicolegys, fixed! Could you also make tests for other APIs and Coinbase/Difficulty fields of accepted blocks? Also, maybe you have some thoughts about RPC API extension (ref. https://github.com/bane-labs/go-ethereum/issues/96)?

nicolegys commented:
> @nicolegys, fixed! Could you also make tests for other APIs and Coinbase/Difficulty fields of accepted blocks? Also, maybe you have some thoughts about RPC API extension (ref. #96)?

Of course! Let me make some tests and then think about extension.

nicolegys commented:
Wonder if it is appropriate to return the `index 0` validator as a primary of genesis block.

**getPrimary**
![1705397548923](https://github.com/bane-labs/go-ethereum/assets/43694095/e8e1aba0-7f9b-48a4-9e5f-4ef1f29420fc)

**getValidators**
![1705398130889](https://github.com/bane-labs/go-ethereum/assets/43694095/50ae45c5-a526-4c66-8745-4f6a90cd3269)


Since miner of `block 0` is zero address.
![1705397970100](https://github.com/bane-labs/go-ethereum/assets/43694095/d073d5f1-5363-4732-88ef-6f2c26a18d86)

AnnaShaleva commented:
> Wonder if it is appropriate to return the index 0 validator as a primary of genesis block.

It's appropriate because the PrimaryIndex of Genesis block is 0, and thus, the 0-th validator from the sorted list of validators should be returned. And for block 1 it's expected to be 1-th validator address to be a primary (if no CVs happens during block 1 acceptance). Also, Coinbase in general shouldn't match the Primary since we've decide that Coinbase is a customizable value that can be set via `miner.etherbase` for every consensus node. It's supposed to be set to Governance contract address in future.

> Since miner of block 0 is zero address

But this is something that may be improved, we need to decide about it. We may set Genesis Coinbase to some address (Governance contract as we've agreed), but then we have to provide it via node's configuration. And then likely to validate `miner.etherbase` against Genesis Coinbase (two values should match). But then we won't be able to "change" `miner.etherbase` in the existing running network without hardfork or some node/dBFT setting extension. From the other hand, we may leave Genesis Coinbase empty which also won't hurt much.

@roman-khimov, what do you think?

roman-khimov commented:
We have to have some primary index for all blocks and zero is always some valid index -> we have zero in genesis, it' OK. See N3 mainnet as well (where primary 0 had not produced this block as well):
```
{
   "result" : {
      "confirmations" : 4744189,
      "version" : 0,
      "merkleroot" : "0x0000000000000000000000000000000000000000000000000000000000000000",
      "time" : 1468595301000,
      "nextblockhash" : "0x5f3fbb43d1e516fe07771e2e873ebc9e2810662401acf603a775aace486220bd",
      "hash" : "0x1f4d1defa46faa5e7b9b8d3f79a06bec777d7c26c4aa5f6f5899a291daa87c15",
      "witnesses" : [
         {
            "invocation" : "",
            "verification" : "EQ=="
         }
      ],
      "nonce" : "000000007C2BAC1D",
      "primary" : 0,
      "index" : 0,
      "size" : 114,
      "previousblockhash" : "0x0000000000000000000000000000000000000000000000000000000000000000",
      "tx" : [],
      "nextconsensus" : "NVg7LjGcUSrgxgjX3zEgqaksfMaiS8Z6e1"
   },
   "id" : 5,
   "jsonrpc" : "2.0"
}
```

There is no real miner for block 0 and there are no fees to distribute -> zero miner address is fine.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -81,6 +81,7 @@ var (
 
 	uncleHash = types.CalcUncleHash(nil) // Always Keccak256(RLP([])) as uncles are meaningless outside of PoW.
 
+	diffStub   = big.NewInt(3) // Block difficulty stub that is used for miner's task preparation.
```
Shouldn't it be zero like in PoS engines? Although Clique uses 2 from what I see...

AnnaShaleva replied:
If it's zero, then sealing tasks will be rejected by Beacon chain.

roman-khimov replied:
OK.

AnnaShaleva committed:
eth: override miner/eth etherbase for dBFT engine assistance

Use `miner.etherbase` setting to specify the dBFT CN account to unlock
and sign accepting blocks.

Override miner's and eth's etherbase with value specified in the dBFT's
Coinbase config to provide proper Coinbase field for accepted blocks.
Block's Coinbase then is the address to send mining reward to.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: implement Coinbase header's verification

Verify Coinbase on PrepareRequest verification level and on block
verification level.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented on consensus/dbft/dbft.go:
```diff
@@ -753,9 +756,10 @@ func (c *DBFT) verifyCascadingFields(chain consensus.ChainHeaderReader, header *
+	// This check forces Coinbase to be the same among all consensus nodes.
+	if header.Coinbase != c.config.Coinbase {
+		return fmt.Errorf("invalid Coinbase: expected %s, got %s", c.config.Coinbase, header.Coinbase)
+	}
```
@roman-khimov, @chenquanyu let's double-check, are we OK with this change? With this code we can't use different Coinbase addresses for different CNs anymore. This address should be shared between all CNs and should be the same within the whole network lifetime.

roman-khimov replied:
> we can't use different Coinbase addresses for different CNs anymore

It's OK if we're managing fee distribution via a contract.

> This address should be shared between all CN

Sure.

> should be the same within the whole network lifetime.

But this can be a problem. Still, if we want to change the contract, likely we can do this with a hardfork.

AnnaShaleva replied:
> It's OK if we're managing fee distribution via a contract.

This is not yet accepted, that's why I've created this conversation.

> Still, if we want to change the contract, likely we can do this with a hardfork.

We can. Or, if needed for NeoX T2, I can add something like NativeUpdateHistory or ValidatorsHistory but for Coinbase:
```
coinbaseHistory:
    0: "0x1212100000000000000000000000000000000001"
    100500: "0x2323200000000000000000000000000000000002"
```

Do we need this right now?

roman-khimov replied:
I don't think it's needed right now, not hard to extend in future.

chenquanyu replied:
> should be the same within the whole network lifetime.

We design the governance contract as upgradable proxy contract, so it's not a critical problem.

AnnaShaleva commented on privnet/four/genesis_template.json:
```diff
@@ -21,7 +21,8 @@
         "0xnode3",
         "0xnode4"
-      ]
+      ],
+      "coinbase": "0x1212100000000000000000000000000000000001"
```
@nicolegys, please, check one more time, now the following problem should be fixed:
```
Now the code requires miner.etherbase to be a local account to sign, so it can't be an external account (like a contract address). Need some changes if we choose to set coinbase by configure miner.etherbase. 
```

nicolegys replied:
Yes, now the beneficiary can be configured by `coinbase`.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -214,11 +214,9 @@ type DBFT struct {
 	recents    *lru.Cache[common.Hash, *Snapshot] // Snapshots for recent block to speed up reorgs
 	signatures *sigLRU                            // Signatures of recent blocks to speed up mining
 
-	proposals map[common.Address]bool // Current list of proposals we are pushing
```
Seems to be mixed into 96fb5e9a8d3e40771a3c38daa7f8edf329849698.

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/api.go:
```diff
@@ -129,34 +108,29 @@ type status struct {
-	if err != nil {
-		return nil, err
+	if n != nil {
+		numBlocks = *n
```
So I can request statistics for 1M blocks? It'll take some time for the node to get it.

AnnaShaleva replied:
Added max constraint.

AnnaShaleva committed:
dbft: remove unused code

Drop snapshot-related code, snapshot mechanism isn't needed for
proper dBFT functioning. Clique voting scheme isn't applicable to
dBFT, thus, remove it.

The final part of #51.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

roman-khimov commented:
Let's wait for @nicolegys to confirm.

nicolegys commented:
`dbft_status` consumes cpu, we may need a param that can be configured via node configuration to limit the number of blocks.

No-api call
![1705547880939](https://github.com/bane-labs/go-ethereum/assets/43694095/2af4f36e-6fd2-4799-9be5-0f5f3ab41a29)

Call `dbft_status` with no param (default: 64), 1 thread, thread delay in 1 ms
![1705548150994](https://github.com/bane-labs/go-ethereum/assets/43694095/5a11a4af-9bc2-4f96-aeb8-6ddc43ecd737)

Call `dbft_status` with param=56000, 1 thread, thread delay in 1 ms
![1705548409622](https://github.com/bane-labs/go-ethereum/assets/43694095/24f2a08e-d2ef-4c6e-8882-1b397ad9f9c9)

nicolegys reviewed (commented):
I think the `coinbase` field should be immutable.

![1705549953206](https://github.com/bane-labs/go-ethereum/assets/43694095/aa7b3b02-86b0-4143-80e6-eaf91f29f538)

 just like if I shut down the geth instance, change the value of `standbyValidators` and then run `geth init`, it will fail.
 
![1705550506217](https://github.com/bane-labs/go-ethereum/assets/43694095/100ea97a-24ad-45c7-9da1-99b6ce62ecb7)

AnnaShaleva commented:
I have a set of fixes for questions posted by @nicolegys, so not ready for the merge yet.

nicolegys commented:
> I think the `coinbase` field should be immutable.
> 
> ![1705549953206](https://private-user-images.githubusercontent.com/43694095/297609419-aa7b3b02-86b0-4143-80e6-eaf91f29f538.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1Njk2NDIsIm5iZiI6MTcwNTU2OTM0MiwicGF0aCI6Ii80MzY5NDA5NS8yOTc2MDk0MTktYWE3YjNiMDItODZiMC00MTQzLTgwZTYtZWFmOTFmMjlmNTM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDA5MTU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2MWJhODA3MjEyMDI0N2M5MGI2MWI0Mzc0NWIxOTNjNjhkYzA1NGY2OGY2OWU1YWE3NjFiY2EzMjg3MmRjZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aZ8OZA2f-16SapEarUWcFvFg9tzPavkJ1DWEjEcgK58)
> 
> just like if I shut down the geth instance, change the value of `standbyValidators` and then run `geth init`, it will fail.
> 
> ![1705550506217](https://private-user-images.githubusercontent.com/43694095/297610610-100ea97a-24ad-45c7-9da1-99b6ce62ecb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1Njk2NDIsIm5iZiI6MTcwNTU2OTM0MiwicGF0aCI6Ii80MzY5NDA5NS8yOTc2MTA2MTAtMTAwZWE5N2EtMjRhZC00NWM3LTlkYTEtOTliNmNlNjJlY2I3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDA5MTU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwNjY0MDRiM2NjMDE3NmY3ZDFlYTFkZDgwZWFhMWY4MTMzNDRhYjE5M2I5OTEzODFjNjI0NDBjYzJhNDEwNGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MY6CsiXc8w8oPaU0Uw4CSeBqPIhc9-e7SAGXq1E012A)

@AnnaShaleva By the way, I've found that the CNs can still reach consensus if all of them change the `coinbase` to the same value and reinit. But If `coinbase` is immutable, it doesn't matter.

AnnaShaleva committed:
dbft: improve logging

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
> @AnnaShaleva By the way, I've found that the CNs can still reach consensus if all of them change the `coinbase` to the same value and reinit.

This is expected result since all dBFT nodes have an agreement on new Coinbase (they don't care what's the old one). But conceptually we need to store the whole Coinbase history for proper old blocks verification, and I'll create a separate issue for that. And for now I've prohibited to change Coinbase at all at the dBFT engine level, take a look at the latest two commits. So now after node reinitialisation with changed Coinbase or StandByValidators we've got an error after node start:
```
anna@kiwi:~/Documents/GitProjects/bane-labs/go-ethereum$ ./build/bin/geth init --datadir privnet/four/node1/ privnet/four/node1/genesis_privnet.json >> privnet/four/node1/geth_init.log 
INFO [01-18|12:27:09.176] Maximum peer count                       ETH=50 LES=0 total=50
INFO [01-18|12:27:09.178] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [01-18|12:27:09.180] Using pebble as the backing database 
INFO [01-18|12:27:09.180] Allocated cache and file handles         database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/chaindata cache=512.00MiB handles=2048
INFO [01-18|12:27:09.180] Opened ancient database                  database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/chaindata/ancient/chain readonly=true
INFO [01-18|12:27:09.180] State scheme set to already existing     scheme=hash
INFO [01-18|12:27:09.181] Set global gas cap                       cap=50,000,000
INFO [01-18|12:27:09.181] Initializing the KZG library             backend=gokzg
INFO [01-18|12:27:09.212] Using pebble as the backing database 
INFO [01-18|12:27:09.212] Allocated cache and file handles         database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/chaindata cache=16.00MiB  handles=16
INFO [01-18|12:27:09.222] Opened ancient database                  database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/chaindata/ancient/chain readonly=false
INFO [01-18|12:27:09.222] State scheme set to already existing     scheme=hash
INFO [01-18|12:27:09.223] Successfully wrote genesis state         database=chaindata hash=98bb61..611263
INFO [01-18|12:27:09.223] Using pebble as the backing database 
INFO [01-18|12:27:09.223] Allocated cache and file handles         database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/lightchaindata cache=16.00MiB  handles=16
INFO [01-18|12:27:09.233] Opened ancient database                  database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/lightchaindata/ancient/chain readonly=false
INFO [01-18|12:27:09.233] State scheme set to already existing     scheme=hash
INFO [01-18|12:27:09.235] Successfully wrote genesis state         database=lightchaindata hash=98bb61..611263
anna@kiwi:~/Documents/GitProjects/bane-labs/go-ethereum$ make privnet_start_four 
Starting nodes...
OK! Check logs in ./privnet/four/<node_dir>/geth_node.log
anna@kiwi:~/Documents/GitProjects/bane-labs/go-ethereum$ cat privnet/four/node1/geth_node.log 
INFO [01-18|12:27:21.352] Enabling metrics collection 
INFO [01-18|12:27:21.352] Maximum peer count                       ETH=50 LES=0 total=50
INFO [01-18|12:27:21.354] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [01-18|12:27:21.356] Using pebble as the backing database 
INFO [01-18|12:27:21.356] Allocated cache and file handles         database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/chaindata cache=512.00MiB handles=2048
INFO [01-18|12:27:21.356] Opened ancient database                  database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/chaindata/ancient/chain readonly=true
INFO [01-18|12:27:21.356] State scheme set to already existing     scheme=hash
INFO [01-18|12:27:21.357] Set global gas cap                       cap=50,000,000
INFO [01-18|12:27:21.357] Initializing the KZG library             backend=gokzg
INFO [01-18|12:27:21.449] Allocated trie memory caches             clean=154.00MiB dirty=256.00MiB
INFO [01-18|12:27:21.449] Using pebble as the backing database 
INFO [01-18|12:27:21.449] Allocated cache and file handles         database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/chaindata cache=512.00MiB handles=2048
INFO [01-18|12:27:21.486] Opened ancient database                  database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/four/node1/geth/chaindata/ancient/chain readonly=false
Fatal: Failed to register the Ethereum service: failed to create dBFT engine: dBFT config has been changed: Coinbase mismatch: expected 0x1212100000000000000000000000000000000001, got 0x1212100000000000000000000000000000000002
```

AnnaShaleva commented:
@nicolegys, I've created a separate issue for RPC dBFT statistics call, see https://github.com/bane-labs/go-ethereum/issues/102, since it's not critical it will be done in a separate PR.

The issue with immutable coinbase is resolved in the current PR since it's a good practice to notify the node's operator about invalid config. But as an enhancement we'll have https://github.com/bane-labs/go-ethereum/issues/101.

So take a look one more time, please, we're going to include this PR into NeoX T2. @roman-khimov @chenquanyu

nicolegys commented on consensus/dbft/api.go:
```diff
@@ -129,34 +113,29 @@ type status struct {
-	snap, err := api.clique.snapshot(api.chain, header.Number.Uint64(), header.Hash(), nil)
-	if err != nil {
-		return nil, err
+	if n != nil && *n <= maxStatisticsPeriod {
```
I think if `n>maxStatisticsPeriod`, n should be set as `maxStatisticsPeriod` instead of `defaultStatisticsPeriod`.

AnnaShaleva replied:
Sure, fixed.

chenquanyu commented on consensus/dbft/config.go:
```diff
@@ -0,0 +1,106 @@
+// A set of Eth database prefixes related to DBFT module.
+var (
+	prefixDBVersion = byte(0x01)
+	prefixDBConfig  = byte(0x02)
```
We stored the dbft config when genesis init, so can we just check the config in `(c *ChainConfig) checkCompatible()`?

AnnaShaleva replied:
Oh, my bad, I haven't noticed the existing verification code. Will fix.

AnnaShaleva replied:
Nice catch!

AnnaShaleva replied:
OK, finally fixed. Geth has quite complicated genesis initialisation process, and the case is that we can't just extend `(c *ChainConfig) checkCompatible()`, because this method should return only special hardfork-related errors that require the node to be reset to some particular height after this error detected. And for some reason we always have genesis block returned from `rawdb.ReadHeadHeader(db)`, there might be some problem with block persist, it shouldn't be like that, will create an issue for this.

And to solve the problem you've mentioned I created a special `(c *ChainConfig) CheckConsensusEngineCompatible`, so now it's forbidden to overwrite an existing genesis on node reinitialisation if consensus engine config is incompatible. Check the latest commit.

AnnaShaleva replied:
Ref. https://github.com/bane-labs/go-ethereum/issues/105 for the head header problem.

nicolegys reviewed (approved):
OK from test side, wait the unresolved coversation to be confirmed.

AnnaShaleva committed:
config: ensure genesis consensus engine config compat on reinit

On node's genesis reinitialisation we need to ensure that consensus engine
configuration is compatible with the previous one. This commit makes the
node operator preserve dBFT config immutable fields.

Ref. https://github.com/bane-labs/go-ethereum/pull/93#pullrequestreview-1828743197.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@nicolegys, take a look one more time please, I've refactored dBFT config compliance check and moved it to `init` level.
